### PR TITLE
[RISCV] Implement RISCVISD::SHL_ADD and move patterns into combine

### PR DESCRIPTION
**Pull request by preames**
This implements a RISCV specific version of the SHL_ADD node proposed in https://github.com/llvm/llvm-project/pull/88791.

If that lands, the infrastructure from this patch should seamlessly switch over the to generic DAG node.  I'm posting this separately because I've run out of useful multiply strength reduction work to do without having a way to represent MUL X, 3/5/9 as a single instruction.

The majority of this change is moving two sets of patterns out of tablgen and into the post-legalize combine.  The major reason for this is that I have an upcoming change which needs to reuse the expansion logic, but it also helps common up some code between zba and the THeadBa variants.

On the test changes, there's a couple major categories:
* We chose a different lowering for mul x, 25.  The new lowering involves one fewer register and the same critical path, so this seems like a win.
* The order of the two multiplies changes in (3,5,9)*(3,5,9) in some cases. I don't believe this matters.
* I'm removing the one use restriction on the multiply.  This restriction doesn't really make sense to me, and the test changes appear positive.

### CODE REVIEW

**Comment by wangpc-pp** [comment]
Sorry for bothering, but I have to say that this account (@pcwang-thead) is not used and I can't access it now. Hopefully its access can be revoked after six months (https://discourse.llvm.org/t/rfc2-new-criteria-for-commit-access/77110), or @tstellar can just remove it.

File: llvm/lib/Target/RISCV/RISCVInstrInfoXTHead.td [hunk 1]
```diff
@@ -581,30 +583,6 @@ def : Pat<(mul (XLenVT GPR:$r), C9LeftShift:$i),
...
-                    (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 2)), 2)>;
-def : Pat<(mul_const_oneuse GPR:$r, (XLenVT 41)),
-          (TH_ADDSL GPR:$r, (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 2)), 3)>;
-def : Pat<(mul_const_oneuse GPR:$r, (XLenVT 73)),
-          (TH_ADDSL GPR:$r, (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 3)), 3)>;
-def : Pat<(mul_const_oneuse GPR:$r, (XLenVT 27)),
-          (TH_ADDSL (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 3)), (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 3)), 1)>;
-def : Pat<(mul_const_oneuse GPR:$r, (XLenVT 45)),
-          (TH_ADDSL (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 3)), (XLenVT (TH_ADDSL GPR:$r, GPR:$r, 3)), 2)>;
-def : Pat<(mul_const_oneuse GPR:$r, (XLenVT 81)),
```
Comment (4vtomat) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXTHead.td:605]: 81 is not in the switch/case in the code above, is it removed accidentally?

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXTHead.td:605]: No.  The switch has "case 91:" which appears to have been a typo.

**Comment by preames** [comment]
ping?

File: llvm/lib/Target/RISCV/RISCVInstrInfoZb.td [hunk 2]
```diff
@@ -26,6 +26,12 @@
 // Operand and SDNode transformation definitions.
 //===----------------------------------------------------------------------===//
 
+def SDTIntShiftAddOp : SDTypeProfile<1, 3, [   // shl_add
+  SDTCisSameAs<0, 1>, SDTCisSameAs<0, 3>, SDTCisInt<0>, SDTCisInt<2>,
+  SDTCisInt<3>
+]>;
+
+def riscv_shl_add    : SDNode<"RISCVISD::SHL_ADD"   , SDTIntShiftAddOp>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:34]: Formatting is weird on this line. I assume it was copied from below but the spacing wasn't updated.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**